jsonnet/kube-prometheus/blackbox-exporter/blackbox-exporter.libsonnet

```diff
@@ -0,0 +1,211 @@
+        },
+        http_post_2xx: {
+          prober: 'http',
+          http: {
+            method: 'POST'
+          }
+        },
+        tcp_connect: {
+          prober: 'tcp'
+        },
+        pop3s_banner: {
+          prober: 'tcp',
+          tcp: {
+            query_response: [
+              { expect: '^+OK' }
+            ],
+            tls: true,
+            tls_config: {
+              insecure_skip_verify: false
+            }
+          }
+        },
+        ssh_banner: {
+          prober: 'tcp',
+          tcp: {
+            query_response: [
+              { expect: '^SSH-2.0-' }
+            ]
+          }
+        },
+        irc_banner: {
+          prober: 'tcp',
+          tcp: {
+            query_response: [
+              { send: 'NICK prober' },
+              { send: 'USER prober prober prober :prober' },
+              { expect: 'PING :([^ ]+)', send: 'PONG ${1}' },
+              { expect: '^:[^ ]+ 001' }
+            ]
+          }
+        },
+      },
+      privileged:
+        local icmpModules = [self.modules[m] for m in std.objectFields(self.modules) if self.modules[m].prober == 'icmp'];
+        std.length(icmpModules) > 0
+    }
+  },
+
+  blackboxExporter+::
+    local bb = $._config.blackboxExporter;
+    {
+      configuration: {
+        apiVersion: 'v1',
+        kind: 'ConfigMap',
+        metadata: {
+          name: 'blackbox-exporter-configuration',
+          namespace: $._config.namespace
+        },
+        data: {
+          'config.yml': std.manifestYamlDoc({ modules: bb.modules })
+        }
+      },
+
+      serviceAccount: {
+        apiVersion: 'v1',
+        kind: 'ServiceAccount',
+        metadata: {
+          name: 'blackbox-exporter',
+          namespace: $._config.namespace,
+        },
+      },
+
+      deployment: {
+        apiVersion: 'apps/v1',
+        kind: 'Deployment',
+        metadata: {
+          name: 'blackbox-exporter',
+          namespace: $._config.namespace,
+          labels: bb.assignLabels,
+        },
+        spec: {
+          replicas: bb.replicas,
+          selector: { matchLabels: bb.matchLabels },
+          template: {
+            metadata: { labels: bb.assignLabels },
+            spec: {
+              containers: [
+                {
+                  name: 'blackbox-exporter',
+                  image: $._config.imageRepos.blackboxExporter + ':' + $._config.versions.blackboxExporter,
+                  ports: [{
+                    name: 'http',
+                    containerPort: bb.port,
+                  }],
+                  resources: {
+                    requests: $._config.resources['blackbox-exporter'].requests,
+                    limits: $._config.resources['blackbox-exporter'].limits
+                  },
+                  securityContext: if bb.privileged then {
+                                     runAsNonRoot: false,
+                                     capabilities: { drop: [ 'ALL' ], add: [ 'NET_RAW'] }
```

Comment:
Why would this be necessary?

Reply:
ICMP probes fail with permission errors if the blackbox exporter process does not have the `NET_RAW` capability.

I first added the capability to the normal unprivileged security context, but that doesn't change the effective capabilities of the processes in the container, just the allowed ones; to make it effective, I'd have to `setcap` the blackbox exporter binary, but that would mean rolling a custom container image, and I don't want to deal with that. The working solution was to run the container as root, but remove all its special privileges and only let it keep `NET_RAW`. (See e.g. prometheus/blackbox_exporter#689)

Although these privileges are quite narrow, I still didn't want to set them this way unless necessary. Folks on secure clusters may need to set up PSP to allow this; if they don't need ICMP probes, we may save them the hassle.